[WFLY-9407] Upgrade Artemis 2.4.0

jmesnil opened the pull request:
[WFLY-9407] Upgrade Artemis 2.4.0
[WFLY-9483] Upgrade Netty to 4.1.16.Final
[WFLY-9484] Upgrade netty-xnio-transport to 0.1.3.Final
[WFLY-9513] Support HA in Messaging JDBC store

This PR contains commits required to upgrade Artemis to 2.4.0. It includes additional component upgrades for components that are only used by Artemis and the `messaging-activemq` subsystem: Netty & netty-xnio-transport.
It also contains [WFLY-9513] Support HA in Messaging JDBC store (that is new in Artemis 2.4.0).

Apart from WFLY-9513 (that I had to do at the same time than the component upgrade), there are no feature changes (Artemis 1.x -> 2.x deals mostly with internal architecture changes).

* bump messaging-activemq subsystem management version to 3.0
* add new 3.0 parser and XSD for messaging-activemq subsystem
* remove JMS-based management API and replace it with Core management
  API
* fix jms-queue & jms-topic :remove operations that must both use the
JMSServerManager to create and destroy the JMS resources
* remove (unused) bindings from JMS<Queue|Topic>Service
* flag properties removed from Artemis 2 as deprecated attributes
* udpate messaging transformers and load them using the org.jboss.as.controller.transform.ExtensionTransformerRegistration
  service loader

JIRA: https://issues.jboss.org/browse/WFLY-9407
JIRA: https://issues.jboss.org/browse/WFLY-9483
JIRA: https://issues.jboss.org/browse/WFLY-9484
JIRA: https://issues.jboss.org/browse/WFLY-9513

jmesnil committed:
[WFLY-9407] Upgrade Artemis 2.4.0

* bump messaging-activemq subsystem to 3.0
* add new 3.0 parser and XSD for messaging-activemq subsystem
* remove JMS-based management API and replace it with Core management
  API
* add "" as the default value for the cluster-connection's address
  attribute to allow any address to be routed to the cluster (Artemis
  2.4.0 no longer prefixes its JMS addresses with jms.queue or
  jms.topic).
* fix jms-queue & jms-topic :remove operations that must both use the
  JMSServerManager to create and destroy the JMS resources
* remove (unused) bindings from JMS<Queue|Topic>Service
* flag properties removed from Artemis 2 as deprecated attributes
* udpate messaging transformers and load them using the org.jboss.as.controller.transform.ExtensionTransformerRegistration
  service loader

JIRA: https://issues.jboss.org/browse/WFLY-9407

jmesnil committed:
[WFLY-9513] Support HA in Messaging JDBC store

* add new attributes to configure HA in Messaging JDBC store:
  * journal-jdbc-lock-acquisition-timeout
  * journal-jdbc-lock-expiration
  * journal-jdbc-lock-renew-period
  * journal-node-manager-store-table
* add new SQL statements in PropertySQLProviderFactory.

JIRA: https://issues.jboss.org/browse/WFLY-9513

jmesnil committed:
[WFLY-9483] Upgrade Netty to 4.1.16.Final

JIRA: https://issues.jboss.org/browse/WFLY-9483

jmesnil committed:
[WFLY-9484] Upgrade netty-xnio-transport to 0.1.3.Final

JIRA: https://issues.jboss.org/browse/WFLY-9484

bstansberry commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java:
```diff
@@ -169,7 +177,7 @@ public static ResourceDescriptionResolver getResourceDescriptionResolver(final b
     public void initialize(ExtensionContext context) {
         final SubsystemRegistration subsystemRegistration = context.registerSubsystem(SUBSYSTEM_NAME, CURRENT_MODEL_VERSION);
-        subsystemRegistration.registerXMLElementWriter(new MessagingSubsystemParser_2_0());
+        subsystemRegistration.registerXMLElementWriter(MessagingSubsystemParser_3_0::new);
```
Please use new MessagingSubsystemParser_3_0().

The method reference doesn't prevent loading of the class, but it does add more metaspace overhead than a constructed instance adds to the heap.

jmesnil replied:
I've no issue with changing that back to instantiating the object.

However, we should update the `registerXMLElementWriter` javadoc:  its `registerXMLElementWriter(XMLElementWriter<SubsystemMarshallingContext> writer)` is `@Deprecated` and recommends to use the variant where we pass a `Supplier` instead.

jmesnil replied:
fixed (by amending the WFLY-9407 commit)

bstansberry replied:
@jmesnil Good point. https://github.com/wildfly/wildfly-core/pull/2963

ctomc replied:
Well I do not agree with this scenario here.
Writers are rarely used as they are only invoked when there is configuration change, which in production systems happens rarely. This why it makes sense to have them created only on demand instead of up-front.

Parsers on other hand, for boot performance it would somewhat make sense to have "current" parser eagerly created.
But if you do that, parser is not GCable.

Maybe it would make sense to do have another version of registerParser(Supplier, boolean) method which would on register invoke get(), and create the parser before boot process would kick in.
This way we would have eager creation on extension init() as well as having them GC-able.

bstansberry commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java:
```diff
@@ -199,14 +207,11 @@ public void initialize(ExtensionContext context) {
     public void initializeParsers(ExtensionParsingContext context) {
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, MessagingSubsystemParser_1_0.NAMESPACE, MessagingSubsystemParser_1_0::new);
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, MessagingSubsystemParser_2_0.NAMESPACE, MessagingSubsystemParser_2_0::new);
+        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, MessagingSubsystemParser_3_0.NAMESPACE, MessagingSubsystemParser_3_0::new);
```
Please use new MessagingSubsystemParser_3_0(), but keep the method reference for the other versions.

This is a bit of black art. When a Supplier is not used, ExtensionRegistry is able to detect use of a PersistentResourceXMLParser and it then triggers a call to getParserDescription(). The object being to have that call happen during extension initialization, which is part of the parallelized part of boot, rather than having it be deferred until parsing, which is single-threaded. Since a getParserDescription() impl typically results in a low of loading of classes from the extension impl, execution of all the static initialization therein, etc, there's a pretty decent boot time gain from getting it out of the single threaded parts of boot.

Using a Supplier for the legacy versions prevents pointless execution of code for those. There's a small metaspace cost though. Tradeoffs. :)


(One of my goals for next-gen declarative subsystem dev is to make getting this stuff right automagic, as this is definitely black art now. I added the ExtensionRegistry trick late in the WF 11 cycle when we were looking for small boot time gains.)

jmesnil replied:
note for later, maybe the API should distinguish between registering the subsystem XML mapping for the current version (and load an instance) vs. legacy XML mappings (where a supplier would prevent unnecessary code execution).

jmesnil replied:
fixed (by amending the WFLY-9407 commit)

bstansberry replied:
@jmesnil My dream is you just register your model and we generate the current parser. Specifically registered parsers would be for old versions only.

bstansberry commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/JMSQueueRemove.java:
```diff
@@ -56,18 +54,16 @@ private JMSQueueRemove() {
-            } catch (Exception e) {
-                throw new OperationFailedException(e);
-            }
+        ServiceController<?> service = context.getServiceRegistry(false).getService(jmsServiceName);
```
This should be getServiceRegistry(true).

It doesn't have a practical effect because this op would have already gotten the exclusive op lock in Stage.RUNTIME, but it's good as a form of illustration to have our code always use 'true' if the code is accessing MSC with the intent to modify runtime state.

jmesnil replied:
fixed (by amending the WFLY-9407 commit).

I misunderstood the `getServiceRegistry(boolean)` javadoc. it says:
> * @param modify {@code true} if the operation may be modifying a service, {@code false} otherwise
     
I understand this as passing `true` if the service *itself* is modified (e.g. calling `setMode`, `{add|remove}Listener`). Maybe we should rephrase: "if the runtime state of a service is modified".

bstansberry replied:
@jmesnil Yeah, that keeps being an issue. https://github.com/wildfly/wildfly-core/pull/2964

bstansberry commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/JMSTopicRemove.java:
```diff
@@ -53,17 +51,16 @@ private JMSTopicRemove() {
-            } catch (Exception e) {
-                throw new OperationFailedException(e);
-            }
+        ServiceController<?> service = context.getServiceRegistry(false).getService(jmsServiceName);
```
getServiceRegistry(true)

jmesnil replied:
fixed (by amending the WFLY-9407 commit).

jmesnil commented:
please do not merge this PR (we have identified a possible regression in the `cluster-connection` resource configuration).

jmesnil committed:
[WFLY-9407] Upgrade Artemis 2.4.0

* add "" as the default value for the cluster-connection's address
  attribute to allow any address to be routed to the cluster (Artemis
  2.4.0 no longer prefixes its JMS addresses with jms.queue or
  jms.topic).
* update XML parser for 1.0 and 2.0 schema to update the
  cluster-connection-address if that was set to "jms" in previous
  configuration files
* update default configuration to change the value of the
  dead-letter-address and expiry-address for the address-settings.

JIRA: https://issues.jboss.org/browse/WFLY-9407

jmesnil commented:
@bstansberry I've added a commit https://github.com/wildfly/wildfly/pull/10641/commits/a622c30abf068666cc110d8f210af4e5cde17182 to fix an issue with cluster connections (I've also reenabled the ClusteredMessagingTestCase that would have caught this issue)

jmesnil commented:
@bstansberry all your change requests have beed addressed.

jmesnil commented:
retest this please

jmesnil commented:
retest this please

jmesnil committed:
[WFLY-9591] Correct attributes corresponding to Artemis address

* If the value of an attribute corresponding to an Artemis address use a
  "jms" prefix, warns the user that it is no longer correct for Artemis
  2.x and use the value without the prefix.
* add a special case if the value is "jms" and use "" (empty string)
  instead.

JIRA: https://issues.jboss.org/browse/WFLY-9591

jmesnil committed:
WIP: backwards compatiblity: support runtime-queue=jms.queue.XXX resource

jmesnil committed:
[WFLY-9590] auto-<create|delete>-jms-queues are not supported

* Artemis 2.4.0 no longer supports this configuration for
 address-settings.
* deprecate the auto-create-jms-queues and auto-delete-jms-queues
* add new attributes auto-create-queues, auto-delete-queues,
  auto-create-addresses, and auto-delete-addresses to comply with
  Artemis 2.4 AddressSetting configuration

JIRA: https://issues.jboss.org/browse/WFLY-9590

jmesnil commented:
retest this please

pferraro commented:
FYI - that transient SingletonPartitionTestCase failure:
https://ci.wildfly.org/viewLog.html?buildId=81981&buildTypeId=WF_PullRequest_Windows#testNameId-6871735680237173899
is addressed by https://github.com/wildfly/wildfly/pull/10654

pferraro commented:
@jmesnil @bstansberry Is there anything else preventing approval/merging of this PR?

bstansberry commented:
@pferraro It is going extensive out-of-band (i.e. non-CI) quality verification and won't be merged until it passes.

jmesnil committed:
fix PropertySQLProviderFactory

jmesnil committed:
add jms prefix to be consistent with Artemis 1.x addressing scheme

jmesnil committed:
upgrade netty-xnio-transport 0.1.4.Final

jmesnil committed:
Update PropertySQLProviderFactory implementation

jmesnil committed:
add jms prefix to be consistent with Artemis 1.x addressing scheme

jmesnil committed:
WIP: revert jms prefix scheme

jmesnil committed:
WIP: revert jms prefix scheme

jmesnil commented:
Closing this PR as regression has been found with Artemis 2.4.0 that must be fixed before Artemis is integrated.
